ansibledriver/service/ansible.py

```diff
@@ -221,127 +228,280 @@ def __init__(self, ansible_properties, request_id, lifecycle, display=None):
                 'id': str(task._uuid)
             },
             'hosts': {}
         }
 
     def v2_playbook_on_play_start(self, play):
-        logger.debug('v2_playbook_on_play_start ok {0}'.format(play))
-        self.results.append(self._new_play(play))
+        """
+        Called when a play begins
+        Note: ONE playbook can have MANY plays
+        """
+        logger.debug('v2_playbook_on_play_start: {0}'.format(play))
+        if self.ansible_properties.log_progress_events:
+            play_name = play.get_name().strip()
+            event = PlayStartedEvent(play_name=play_name)
+            self.event_logger.add(event)
+        self.plays.append(self._new_play(play))
 
     def v2_playbook_on_task_start(self, task, is_conditional):
-        logger.debug('v2_playbook_on_task_start ok {0} {1}'.format(task, is_conditional))
+        """
+        Called when a task starts 
+        Note: even if a task is going to run on multiple hosts, this function is only called ONCE
+        """
+        logger.debug('v2_playbook_on_task_start: {0} (is_conditional={1})'.format(task, is_conditional))
+        self._log_task_start(task)
 
     def v2_playbook_on_handler_task_start(self, task):
-        logger.debug('v2_playbook_on_handler_task_start ok {0}'.format(task))
+        logger.debug('v2_playbook_on_handler_task_start: {0}'.format(task))
+        self._log_task_start(task, prefix='Handler/')
+
+    def _log_task_start(self, task, prefix=None):
+        if self.ansible_properties.log_progress_events:
+            if prefix is None:
+              prefix = ''
+            task_name = '{0}{1}'.format(prefix, task.get_name().strip())
+            event = TaskStartedEvent(task_name=task_name)
+            if not task.no_log:
+              # Include args if the task has not been configured with the no_log option
+              for arg_name, arg_value in task.args.items():
+                event.args[str(arg_name)] = str(arg_value)
+            else:
+              event.args_hidden = True
+            self.event_logger.add(event)
 
     def v2_playbook_on_stats(self, stats):
-        """Display info about playbook statistics"""
-
-        hosts = sorted(stats.processed.keys())
-
-        summary = {}
-        for h in hosts:
-            s = stats.summarize(h)
-            summary[h] = s
-
-        output = {
-            'plays': self.results,
-            'stats': summary
-        }
-
-        logger.debug('v2_playbook_on_stats {0}'.format(json.dumps(output, indent=4, sort_keys=True)))
+        """
+        Called at the end of playbook execution (even in failure)
+        """
+        logger.debug('v2_playbook_on_stats: {0}'.format(stats))
+        if self.ansible_properties.log_progress_events:
+            hosts = sorted(stats.processed.keys())
+            host_stats = {}
+            for h in hosts:
+                s = stats.summarize(h)
+                host_stats[h] = s
+            
```

Comment:
minor: can't you just assign directly to host_stats[h]?

Reply:
Fixed the assign. 
My understanding is it's always != None. Got that from looking at the default Ansible callback (that prints the familiar output to the console when using the CLI) at https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/default.py